File: tollfree-edit-window-expiring/node/twilio-tollfree-edit-window.mjs
/**
 * Flag rejected toll-free verifications whose edit window is about to close.
 *
 * Read only. GET requests and nothing else: give this an API Key with read
 * access rather than the account auth token. The correction needs a human who
 * has read the rejection reasons; this only makes sure they still can.
 */
const MSG = 'https://messaging.twilio.com/v1';

const REJECTED = 'TWILIO_REJECTED';

/** Parse a messaging v1 ISO 8601 timestamp. Pure. Returns a Date or null. */
export function parseTime(value) {
  const text = String(value ?? '').trim();
  if (!text) return null;
  const t = Date.parse(text);
  return Number.isNaN(t) ? null : new Date(t);
}

/** Hours until the edit window closes. Negative once it has passed. */
export function hoursLeft(editExpiration, now) {
  const expires = parseTime(editExpiration);
  if (expires === null || !now) return null;
  return (expires.getTime() - now.getTime()) / 3600000;
}

/**
 * Classify one rejected toll-free verification against its edit window. `hours`
 * is the time remaining, or null; taking it as an argument keeps the clock out
 * of the classifier. Nothing here reads the rejection reasons. Pure.
 * Returns [state, detail].
 */
export function verdict(record, hours, horizonHours = 72.0) {
  if (!record) return ['no-record', 'no verification to read.'];

  const status = String(record.status ?? '').toUpperCase();
  if (status !== REJECTED) {
    return ['not-rejected',
      `status is ${status || 'unset'}: there is no edit window on a record ` +
      'that has not been rejected.'];
  }

  const allowed = record.edit_allowed;
  if (allowed === null || allowed === undefined) {
    return ['edit-allowed-unset',
      'rejected, and edit_allowed is absent from the response. That is not the ' +
      'same as false: nothing has been learned about the window, so do not ' +
      'file a fresh submission on this alone.'];
  }

  if (!allowed) {
    return ['no-edit-window',
      'rejected with edit_allowed false. The in-place correction was never on ' +
      'offer here, so a fresh submission is the only path and there is no ' +
      'deadline to race.'];
  }

  if (hours === null || hours === undefined) {
    return ['expiration-unreadable',
      'rejected with edit_allowed true, and edit_expiration could not be ' +
      'parsed. Treat the window as closing and correct now.'];
  }

  if (hours <= 0) {
    return ['window-lapsed',
      `edit_expiration passed ${Math.abs(hours).toFixed(0)} hours ago while ` +
      'edit_allowed still reads true. The timestamp is what the platform ' +
      'enforces, so expect the correction to be refused and plan on a fresh ' +
      'submission.'];
  }

  if (hours <= horizonHours) {
    return ['closing',
      `${hours.toFixed(0)} hours left on the edit window. After that the ` +
      'in-place correction is gone and the only route is a fresh submission, ' +
      'back of the review queue.'];
  }

  return ['open',
    `${hours.toFixed(0)} hours left on the edit window, outside the ` +
    `${horizonHours.toFixed(0)} hour horizon.`];
}

function authHeader(key, secret) {
  return `Basic ${Buffer.from(`${key}:${secret}`).toString('base64')}`;
}

async function get(auth, url, params = {}) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  const res = await fetch(u, { headers: { Authorization: auth } });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from Twilio: check TWILIO_ACCOUNT_SID and ` +
                    'that the API key belongs to that account with read access');
  }
  if (!res.ok) throw new Error(`${res.status} from ${u.pathname}`);
  return res.json();
}

export async function listV1(auth, url, key, limit = 1000, params = {}) {
  const out = [];
  let next = url;
  let first = params;
  while (next && out.length < limit) {
    const page = await get(auth, next, { PageSize: 50, ...first });
    out.push(...(page[key] ?? []));
    next = page.meta?.next_page_url ?? null;
    first = {};
  }
  return out.slice(0, limit);
}

async function main() {
  const account = (process.env.TWILIO_ACCOUNT_SID || "dummy-twilio-account-sid");
  const key = (process.env.TWILIO_API_KEY || "dummy-twilio-api-key");
  const secret = (process.env.TWILIO_API_SECRET || "dummy-twilio-api-secret");
  if (!account || !key || !secret) {
    console.error('set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET ' +
                  '(an API Key with read access, not the auth token)');
    process.exitCode = 2;
    return;
  }
  const auth = authHeader(key, secret);
  const flag = process.argv.indexOf('--horizon-hours');
  const horizon = flag >= 0 ? Number(process.argv[flag + 1]) : 72.0;
  const now = new Date();

  const records = await listV1(auth, `${MSG}/Tollfree/Verifications`,
                               'verifications', 500, { Status: REJECTED });
  if (records.length === 0) {
    console.log('no rejected toll-free verifications on this account');
    return;
  }

  let bad = 0;
  for (const rec of records) {
    const hours = hoursLeft(rec.edit_expiration, now);
    const [state, detail] = verdict(rec, hours, horizon);
    const name = rec.tollfree_phone_number_sid ?? rec.sid ?? 'record';
    const line = `${state.padEnd(22)} ${name}  ${detail}`;
    if (state === 'open' || state === 'no-edit-window') { console.log(line); continue; }
    bad += 1;
    console.warn(line);
    if (state === 'closing' || state === 'expiration-unreadable') {
      console.warn(`  repair: correct the named fields on ${MSG}/Tollfree/` +
                   `Verifications/${rec.sid ?? '{Sid}'} before ` +
                   `${rec.edit_expiration ?? 'the expiration'}, then resubmit. ` +
                   'Console: Phone Numbers, Manage, Active numbers, Regulatory ' +
                   'Information, edit and resubmit');
    } else if (state === 'window-lapsed') {
      console.warn('  repair: file a fresh verification for this number and ' +
                   'expect the full review time; the in-place edit is no longer ' +
                   'available');
    }
  }

  console.log(`${records.length} rejected verification(s), ${bad} closing inside ` +
              `${horizon} hours`);
  process.exitCode = bad ? 1 : 0;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing credentials and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}

File: tollfree-edit-window-expiring/python/twilio_tollfree_edit_window.py
"""Flag rejected toll-free verifications whose edit window is about to close.

Read only. GET requests and nothing else: give this an API Key with read access
rather than the account auth token. The correction itself needs a human who has
read the rejection reasons; this script only makes sure they still can.
"""
import argparse
import datetime
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("twilio_tollfree_edit_window")

MSG = "https://messaging.twilio.com/v1"

REJECTED = "TWILIO_REJECTED"


def parse_time(value):
    """Parse a messaging v1 timestamp. Pure.

    These come back as ISO 8601 with a trailing Z, which
    datetime.fromisoformat did not accept before Python 3.11.
    """
    text = str(value or "").strip()
    if not text:
        return None
    if text.endswith("Z"):
        text = text[:-1] + "+00:00"
    try:
        return datetime.datetime.fromisoformat(text)
    except ValueError:
        return None


def hours_left(edit_expiration, now):
    """Hours until the edit window closes. Negative once it has passed."""
    expires = parse_time(edit_expiration)
    if expires is None or now is None:
        return None
    return (expires - now).total_seconds() / 3600.0


def verdict(record, hours, horizon_hours=72.0):
    """Classify one rejected toll-free verification against its edit window.

    `hours` is the time remaining, or None. Taking it as an argument keeps the
    clock out of the classifier. Nothing here reads the rejection reasons: what
    can be corrected is a separate question from how long there is to do it.
    Returns (state, detail).
    """
    if not record:
        return ("no-record", "no verification to read.")

    status = str(record.get("status") or "").upper()
    if status != REJECTED:
        return ("not-rejected",
                "status is %s: there is no edit window on a record that has "
                "not been rejected." % (status or "unset"))

    allowed = record.get("edit_allowed")
    if allowed is None:
        return ("edit-allowed-unset",
                "rejected, and edit_allowed is absent from the response. That "
                "is not the same as false: nothing has been learned about the "
                "window, so do not file a fresh submission on this alone.")

    if not allowed:
        return ("no-edit-window",
                "rejected with edit_allowed false. The in-place correction was "
                "never on offer here, so a fresh submission is the only path "
                "and there is no deadline to race.")

    if hours is None:
        return ("expiration-unreadable",
                "rejected with edit_allowed true, and edit_expiration could "
                "not be parsed. Treat the window as closing and correct now.")

    if hours <= 0:
        return ("window-lapsed",
                "edit_expiration passed %.0f hours ago while edit_allowed "
                "still reads true. The timestamp is what the platform "
                "enforces, so expect the correction to be refused and plan on "
                "a fresh submission." % abs(hours))

    if hours <= horizon_hours:
        return ("closing",
                "%.0f hours left on the edit window. After that the in-place "
                "correction is gone and the only route is a fresh submission, "
                "back of the review queue." % hours)

    return ("open",
            "%.0f hours left on the edit window, outside the %.0f hour horizon."
            % (hours, horizon_hours))


def get(session, url, **params):
    r = session.get(url, params=params, timeout=30)
    if r.status_code in (401, 403):
        raise SystemExit("%d from Twilio: check TWILIO_ACCOUNT_SID and that the "
                         "API key belongs to that account with read access"
                         % r.status_code)
    r.raise_for_status()
    return r.json()


def list_v1(session, url, key, limit=1000, **params):
    """Page a messaging.twilio.com list. meta.next_page_url is absolute."""
    out = []
    while url and len(out) < limit:
        page = get(session, url, PageSize=50, **params)
        out.extend(page.get(key, []))
        url = (page.get("meta") or {}).get("next_page_url")
        params = {}
    return out[:limit]


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--horizon-hours", type=float, default=72.0,
                    help="how close to the deadline counts as a finding")
    ap.add_argument("--max-records", type=int, default=500)
    args = ap.parse_args()

    account = os.environ.get("TWILIO_ACCOUNT_SID")
    key = os.environ.get("TWILIO_API_KEY")
    secret = os.environ.get("TWILIO_API_SECRET")
    if not (account and key and secret):
        log.error("set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET "
                  "(an API Key with read access, not the auth token)")
        return 2

    session = requests.Session()
    session.auth = (key, secret)
    now = datetime.datetime.now(datetime.timezone.utc)

    records = list_v1(session, MSG + "/Tollfree/Verifications",
                      "verifications", args.max_records, Status=REJECTED)
    if not records:
        log.info("no rejected toll-free verifications on this account")
        return 0

    bad = 0
    for rec in records:
        hours = hours_left(rec.get("edit_expiration"), now)
        state, detail = verdict(rec, hours, args.horizon_hours)
        name = rec.get("tollfree_phone_number_sid") or rec.get("sid") or "record"
        line = "%-22s %s  %s" % (state, name, detail)
        if state in ("open", "no-edit-window"):
            log.info(line)
            continue
        bad += 1
        log.warning(line)
        if state in ("closing", "expiration-unreadable"):
            log.warning("  repair: correct the named fields on %s/Tollfree/"
                        "Verifications/%s before %s, then resubmit. Console: "
                        "Phone Numbers, Manage, Active numbers, Regulatory "
                        "Information, edit and resubmit", MSG,
                        rec.get("sid", "{Sid}"),
                        rec.get("edit_expiration", "the expiration"))
        elif state == "window-lapsed":
            log.warning("  repair: file a fresh verification for this number "
                        "and expect the full review time; the in-place edit is "
                        "no longer available")

    log.info("%d rejected verification(s), %d closing inside %.0f hours",
             len(records), bad, args.horizon_hours)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())
